tests/local.rs

```diff
@@ -0,0 +1,264 @@
+                caps.insert("moz:firefoxOptions".to_string(), opts.clone());
+                Client::with_capabilities("http://localhost:4444", caps)
+            },
+            "chrome" => {
+                let mut caps = serde_json::map::Map::new();
+                let opts = serde_json::json!({
+                    "args": ["--headless", "--disable-gpu", "--no-sandbox", "--disable-dev-shm-usage"],
+                    "binary":
+                        if std::path::Path::new("/usr/bin/chromium-browser").exists() {
+                            // on Ubuntu, it's called chromium-browser
+                            "/usr/bin/chromium-browser"
+                        } else if std::path::Path::new("/Applications/Google Chrome.app/Contents/MacOS/Google Chrome").exists() {
+                            // macOS
+                            "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
+                        } else {
+                            // elsewhere, it's just called chromium
+                            "/usr/bin/chromium"
+                        }
+                });
+                caps.insert("goog:chromeOptions".to_string(), opts.clone());
+
+                Client::with_capabilities("http://localhost:9515", caps)
+            },
+            browser => unimplemented!("unsupported browser backend {}", browser),
+        };
+
+        // we'll need the session_id from the thread
+        // NOTE: even if it panics, so can't just return it
+        let session_id = Arc::new(Mutex::new(None));
+
+        // run test in its own thread to catch panics
+        let sid = session_id.clone();
+        let success = match thread::spawn(move || {
+            let mut rt = tokio::runtime::Builder::new().enable_all().basic_scheduler().build().unwrap();
+            let mut c = rt.block_on(c).expect("failed to construct test client");
+            *sid.lock().unwrap() = rt.block_on(c.session_id()).unwrap();
+            let x = rt.block_on($f(c, port));
+            drop(rt);
+            x
+        })
+        .join()
+        {
+            Ok(Ok(_)) => true,
+            Ok(Err(e)) => {
+                eprintln!("test future failed to resolve: {:?}", e);
+                false
+            }
+            Err(e) => {
+                if let Some(e) = e.downcast_ref::<error::CmdError>() {
+                    eprintln!("test future panicked: {:?}", e);
+                } else if let Some(e) = e.downcast_ref::<error::NewSessionError>() {
+                    eprintln!("test future panicked: {:?}", e);
+                } else {
+                    eprintln!("test future panicked; an assertion probably failed");
+                }
+                false
+            }
+        };
+
+        assert!(success);
+    }};
+}
+
+lazy_static::lazy_static! {
+    static ref PORT_COUNTER: std::sync::atomic::AtomicU16 = std::sync::atomic::AtomicU16::new(8000);
+}
+
+
+fn setup_server() -> u16 {
+    let port: u16;
+    loop {
```

Comment:
Why do this rather than just bind a `TcpListener` without specifying a port, and then extracting the auto-selected port from `listener.local_addr()`?

Reply:
```
    let port = TcpListener::bind("127.0.0.1").unwrap().local_addr().unwrap().port();
```
I just tried the above, and couldn't get it to work.
I'm inclined to leave it as is.

Reply:
I would strongly prefer that we don't do port probing. Letting binding take care of port allocation is the standard way to do this, and much less error-prone. The trick here is probably going to be using [`warp::Server::bind_ephemeral`](https://docs.rs/warp/0.2.2/warp/struct.Server.html#method.bind_ephemeral), pass it a `SocketAddr` with the port set to `0`, and then the `SocketAddr` you get back has the port to use set.

Reply:
I'm _really_ struggling with this approach. 
For whatever reason, starting the server in this manner causes it to not actually come up.
Usually this results in the first test failing to change the url to the expected address, because the server isn't present.

I've tried spinning for ~10 seconds with a http client to see if the server comes up, but it doesn't seem to. (firefox tests will fail if they can't get a valid response, while chrome tests seem to be ok with that. )

This is far more error prone than the former method, at least for the use case of short-lived test servers.

Reply:
That is because you are constructing a `TcpListener`, getting its address, and then passing the address to `warp`. Instead, _just_ use warp's `bind_ephemeral` and pass 0 as the port, and then just extract the port number from the return value of `bind_ephemeral` :) You should not need any retries.

Reply:
Thanks. The latest change greatly simplified the server setup, but as far as I'm aware, it still doesn't work. Hopefully its easier to spot an error now.